### ISSUE TITLE
No need to div `marginal_prob_std(t)` in the output part of the model?

### ISSUE THREAD

**Issue by TrayfourYu** [ZhengYinan-AIR/Diffusion-Planner#62]
Hi, thanks for your great job! @ZhengYinan-AIR 
And here is my questions about 'score' which is the output of the model. 
In the `decoder`
`if self._model_type == "score":
     return x / (self.marginal_prob_std(t)[:, None, None] + 1e-6)`
And in the `loss`:
`mean, std = marginal_prob(all_gt[..., 1:, :], t)
...
if model_type == "score":
        dpm_loss = torch.sum((score * std + z)**2, dim=-1)`
Refrence to `VPSDE_linear`:
`std `from `marginal_prob `equal to `std `from `marginal_prob_std`
Plus, by the SDE thoery, we have:
`z=−sqrt(1−α(t))⋅s(θ)`
`std `in code is equal to `sqrt(1−α(t))`, and s(θ) is the score function which is supposed to be the prediction value of the model. 
Relative to loss, it pushs `score` in code to the `s(θ)`, which is just we wanted.
`score * std + z`
So, its no need to div `marginal_prob_std(t)` in the output part of the model?

By the way , the formulation in paper may missing a sqrt.

<img width="652" height="65" alt="Image" src="https://github.com/user-attachments/assets/950ab165-c2c9-4e2d-84ef-6abd6d341f2b" />

**Comment by donkehuang** [comment]
@TrayfourYu  Hi, i'm confused with the training process. I noticed that the decoder outputs x_0 directly and then the loss was calculated with gt directly.  The noise function \mu_\theta  was updated by the one step score function. I can't understand why it worked?  Do u know something about it?

**Comment by TrayfourYu** [comment]
@donkehuang hi！what do you mean by “The noise function \mu_\theta was updated by the one step score function.”

**Comment by donkehuang** [comment]
> step

I used to wonder why the diffusion planner can achieve good results by only predicting x0, not the noise.

**Comment by TrayfourYu** [comment]
> > step
> 
> I used to wonder why the diffusion planner can achieve good results by only predicting x0, not the noise.

Its the same thing, bro. x_0 will convert to noise too. Just have a look at the code here.
`elif model_type == "x_start":`
`    alpha_t, sigma_t = noise_schedule.marginal_alpha(t_continuous), noise_schedule.marginal_std(t_continuous)`
`    return (x - expand_dims(alpha_t, x.dim()) * output) / expand_dims(sigma_t, x.dim())`
Which is the impliment of the formulation:
`x_t = alpha_t*x_0 + sigma_t*noise`
So,
`noise = (x_t -  alpha_t*x_0) / sigma_t `
(alpha_t is following the define in the code)

**Comment by HerrSchnapps** [comment]
> > > 步
> > 
> > 
> > 我曾经想知道为什么扩散规划器可以通过只预测 x0 而不是噪声来获得良好的结果。
> 
> 这是一回事，兄弟x_0也会转换为噪音。只需查看此处的代码即可。 这是公式的隐含：所以， （alpha_t遵循代码中的定义）`elif model_type == "x_start":`` alpha_t, sigma_t = noise_schedule.marginal_alpha(t_continuous), noise_schedule.marginal_std(t_continuous)`` return (x - expand_dims(alpha_t, x.dim()) * output) / expand_dims(sigma_t, x.dim())``x_t = alpha_t*x_0 + sigma_t*noise``noise = (x_t - alpha_t*x_0) / sigma_t `

I think the mode=="score" is the diffusion model, rather "x_start" ,right?

**Comment by TrayfourYu** [comment]
> > > > 步
> > > 
> > > 
> > > 我曾经想知道为什么扩散规划器可以通过只预测 x0 而不是噪声来获得良好的结果。
> > 
> > 
> > 这是一回事，兄弟x_0也会转换为噪音。只需查看此处的代码即可。 这是公式的隐含：所以， （alpha_t遵循代码中的定义）`` elif model_type == "x_start":`` alpha_t, sigma_t = noise_schedule.marginal_alpha(t_continuous), noise_schedule.marginal_std(t_continuous)`` return (x - expand_dims(alpha_t, x.dim()) * output) / expand_dims(sigma_t, x.dim())``x_t = alpha_t*x_0 + sigma_t*noise``noise = (x_t - alpha_t*x_0) / sigma_t  ``
> 
> I think the mode=="score" is the diffusion model, rather "x_start" ,right?

Both of them are diffusion based model. `x_start` , `noise` and `score function` can convert to each other following the diffusion theory.

**Comment by HerrSchnapps** [comment]
<img width="1148" height="836" alt="Image" src="https://github.com/user-attachments/assets/697ac1f0-da94-4f4d-9191-1d84a018c638" />


<img width="1144" height="799" alt="Image" src="https://github.com/user-attachments/assets/c6d4e70a-6933-4d4a-af08-986caba35b2b" />


<img width="718" height="181" alt="Image" src="https://github.com/user-attachments/assets/62e2f6f1-c666-4eeb-8738-89013c09d628" />


It seems only the "score" mode is related to noise, the "x_start" is the normal supervised network, which is composed of several layers of mlp or attention

**Comment by smilebetterworld** [comment]
> Hi, thanks for your great job! [@ZhengYinan-AIR](https://github.com/ZhengYinan-AIR) And here is my questions about 'score' which is the output of the model. In the `decoder` `if self._model_type == "score": return x / (self.marginal_prob_std(t)[:, None, None] + 1e-6)` And in the `loss`: `mean, std = marginal_prob(all_gt[..., 1:, :], t) ... if model_type == "score": dpm_loss = torch.sum((score * std + z)**2, dim=-1)` Refrence to `VPSDE_linear`: `std `from `marginal_prob `equal to `std `from `marginal_prob_std` Plus, by the SDE thoery, we have: `z=−sqrt(1−α(t))⋅s(θ)` `std `in code is equal to `sqrt(1−α(t))`, and s(θ) is the score function which is supposed to be the prediction value of the model. Relative to loss, it pushs `score` in code to the `s(θ)`, which is just we wanted. `score * std + z` So, its no need to div `marginal_prob_std(t)` in the output part of the model?
> 
> By the way , the formulation in paper may missing a sqrt.
> 
> <img alt="Image" width="652" height="65" src="https://private-user-images.githubusercontent.com/50270077/468539123-950ab165-c2c9-4e2d-84ef-6abd6d341f2b.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTYyNjY2NjUsIm5iZiI6MTc1NjI2NjM2NSwicGF0aCI6Ii81MDI3MDA3Ny80Njg1MzkxMjMtOTUwYWIxNjUtYzJjOS00ZTJkLTg0ZWYtNmFiZDZkMzQxZjJiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODI3VDAzNDYwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE0Y2VkYzM2ODJjM2UxNzAwNDM1ZDAwMTAyNWRkNzlmNjQ2NjBiODEzMjIxNzJjMDI4NTYwMTUwNDgxZDc1OWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.FWr8JQngjIc2gi8qGl3zTAmm-NqUBa-xE-79V18dRkE">

@TrayfourYu ， I think the network predict noise， so it divide marginal_prob_std(t) in the output part of the model so that the final output becomes score

**Comment by TrayfourYu** [comment]
> [@TrayfourYu](https://github.com/TrayfourYu) ， I think the network predict noise， so it divide marginal_prob_std(t) in the output part of the model so that the final output becomes score

@smilebetterworld If the final output is `score `as u expected, why the loss is the diff between `score` and `-(z/std)`?  (by the way, `-(z/std)` is the gt score)
```
if model_type == "score": 
    dpm_loss = torch.sum((score * std + z)**2, dim=-1)
```

**Comment by smilebetterworld** [comment]
> > [@TrayfourYu](https://github.com/TrayfourYu) ， I think the network predict noise， so it divide marginal_prob_std(t) in the output part of the model so that the final output becomes score
> 
> [@smilebetterworld](https://github.com/smilebetterworld) If the final output is `score `as u expected, why the loss is the diff between `score` and `-(z/std)`? (by the way, `-(z/std)` is the gt score)
> 
> ```
> if model_type == "score": 
>     dpm_loss = torch.sum((score * std + z)**2, dim=-1)
> ```

@TrayfourYu , the loss is (score * std + z) which is reasonable, because score * std becomes predicted noise which should be calculated with true noise to get diff,  I think.
